realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProgressListenerTests.java

```diff
@@ -0,0 +1,332 @@
+        SyncConfiguration config = createSyncConfig();
+        Realm realm = Realm.getInstance(config);
+        writeSampleData(realm);
+
+        SyncSession session = SyncManager.getSession(config);
+        session.addUploadProgressListener(ProgressMode.CURRENT_CHANGES, new ProgressListener() {
+            @Override
+            public void onChange(Progress progress) {
+                if (progress.isTransferComplete()) {
+                    assertTransferComplete(progress);
+                    allChangeUploaded.countDown();
+                }
+            }
+        });
+
+        TestHelper.awaitOrFail(allChangeUploaded);
+        realm.close();
+
+    }
+
+    @Test
+    public void uploadProgressListener_indefinetely() {
+        final AtomicInteger transferCompleted = new AtomicInteger(0);
+        final CountDownLatch testDone = new CountDownLatch(1);
+        final SyncConfiguration config = createSyncConfig();
+        Realm realm = Realm.getInstance(config);
+
+        writeSampleData(realm); // Write first batch of sample data
+        SyncSession session = SyncManager.getSession(config);
+        session.addUploadProgressListener(ProgressMode.INDEFINETELY, new ProgressListener() {
+            @Override
+            public void onChange(Progress progress) {
+                if (progress.isTransferComplete()) {
+                    switch(transferCompleted.incrementAndGet()) {
+                        case 1:
+                            Realm realm = Realm.getInstance(config);
+                            writeSampleData(realm);
+                            realm.close();
+                            break;
+                        case 2:
+                            assertTransferComplete(progress);
+                            testDone.countDown();
+                            break;
+                        default:
+                            fail("Unsupported number of transfers completed: " + transferCompleted.get());
+                    }
+                }
+            }
+        });
+
+        TestHelper.awaitOrFail(testDone);
+        realm.close();
+    }
+
+    @Test
+    public void addListenerInsideCallback() {
+        final CountDownLatch allChangeUploaded = new CountDownLatch(1);
+        final SyncConfiguration config = createSyncConfig();
+        Realm realm = Realm.getInstance(config);
+        writeSampleData(realm);
+
+        final SyncSession session = SyncManager.getSession(config);
+        session.addUploadProgressListener(ProgressMode.CURRENT_CHANGES, new ProgressListener() {
+            @Override
+            public void onChange(Progress progress) {
+                if (progress.isTransferComplete()) {
+                    Realm realm = Realm.getInstance(config);
+                    writeSampleData(realm);
+                    realm.close();
+                    session.addUploadProgressListener(ProgressMode.CURRENT_CHANGES, new ProgressListener() {
```

Comment:
No, because when you add a listener with CURRENT_CHANGES, it will record the number of changes when you add it. So adding it before you write, will just make it reported completed straight away because nothing was written yet.